game-app/game-core/src/main/java/games/strategy/triplea/util/UnitSeparator.java

```diff
@@ -48,36 +48,80 @@ public static class SeparatorCategories {
-   * territory 3. Within each of those groups sort the units by XML order in UnitList
+   * Finds unit categories from units of the <code>Territory</code>, removes not displayed ones
+   * according to <code>MapData</code> and then sorts them
    */
   public static List<UnitCategory> getSortedUnitCategories(
       final Territory t, final MapData mapData) {
-    final GameState data = t.getData();
     final List<UnitCategory> categories = new ArrayList<>(UnitSeparator.categorize(t.getUnits()));
     categories.removeIf(uc -> !mapData.shouldDrawUnit(uc.getType().getName()));
-    final List<UnitType> xmlUnitTypes = new ArrayList<>(data.getUnitTypeList().getAllUnitTypes());
-    categories.sort(
-        Comparator.comparing(
-                UnitCategory::getOwner,
-                Comparator.comparing((final GamePlayer p) -> !p.equals(t.getOwner()))
-                    .thenComparing(p -> Matches.isAtWar(p).test(t.getOwner()))
-                    .thenComparing(data.getPlayerList().getPlayers()::indexOf))
-            .thenComparing(uc -> Matches.unitTypeCanMove(uc.getOwner()).test(uc.getType()))
-            .thenComparing(
-                UnitCategory::getType,
-                Comparator.comparing(
-                        (final UnitType ut) ->
-                            !Matches.unitTypeCanNotMoveDuringCombatMove().test(ut))
-                    .thenComparing(ut -> !Matches.unitTypeIsSea().test(ut))
-                    .thenComparing(ut -> !(t.isWater() && Matches.unitTypeIsAir().test(ut)))
-                    .thenComparing(ut -> !Matches.unitTypeIsLand().test(ut))
-                    .thenComparing(xmlUnitTypes::indexOf)));
+    categories.sort(getComparatorUnitCategories(t));
     return categories;
   }
 
+  /** Returns <code>Comparator</code> for unit categories of a <code>Territory</code> */
+  public static Comparator<UnitCategory> getComparatorUnitCategories(
+      final Territory t, final GamePlayer currentPlayer) {
+    final GameData gameData = t.getData();
+    return getComparatorUnitCategories(t, gameData, currentPlayer);
+  }
+
+  /** Returns <code>Comparator</code> for unit categories of a <code>Territory</code> */
+  public static Comparator<UnitCategory> getComparatorUnitCategories(final Territory t) {
+    final GameData gameData = t.getData();
+    return getComparatorUnitCategories(t, gameData, gameData.getHistory().getCurrentPlayer());
+  }
+
+  /** Returns <code>Comparator</code> for unit categories with current <code>GameData</code> */
+  public static Comparator<UnitCategory> getComparatorUnitCategories(final GameData gameData) {
+    return getComparatorUnitCategories(null, gameData, gameData.getHistory().getCurrentPlayer());
+  }
+
+  /** Returns <code>Comparator</code> for unit categories of a <code>Territory</code> */
+  private static Comparator<UnitCategory> getComparatorUnitCategories(
+      final Territory t, final GameData gameData, final GamePlayer currentPlayer) {
```

Comment:
I notice methods calling this one get `GameData` from `Territory`, any reason to not do that here and remove the `GameData` parameter?

Reply:
I am going to create a fix to change `Territory `to `Optional<Territory>`.

Reply:
Cool, though - at a style level consider having multiple method signatures instead. `Optional` is a "bad" type for a method parameter.  

Though, maybe we can avoid the dependency on territory? Is the ordering change potentially that significant considering territory? Is that important? Second, it's an interesting question if it's more valuable to have a fully consistent ordering regardless of even territory. Thus, perhaps it's worthwhile to see if we can simplify the comparator to not depend on territory (while at it, maybe player too)

Reply:
I think we should enable the sorting to allow for some criteria to be optional that should still be considered and territory properties like owner or isWater seem plausble criteria as well as gamePlayer with whether its the same as the unit owner.
For me this comes as a cost of having a general sorting logic that it should incorporate several situations where some criteria might not be present.

Reply:
> properties like owner or isWater seem plausble criteria as well as gamePlayer with whether its the same as the unit owner.

Are these plausible criteria or required? For either case, if you would, can you walk me through some scenarios you considering where player or territory would change the Unit-Category ordering?

Also to consider, we can have a method give a generic ordering of unit category, and then the places that do care about player or territory, they can do their one-off logic locally. That is to say, there is flexibility; just because the top-most level method is not super-flexible, doesn't mean we can't work that flexibility in where needed.

Reply:
Second thought: just please be sure to remove any method signatures that are not used if any.

Minimizing inputs is a way to have a more consistent sort ordering. We probably don't need to draw out the convo any further than there.